### ISSUE TITLE
Consider removing the explicit `Env` argument from `contractimpl` functions

### ISSUE THREAD

**Issue by tomerweller** [stellar/rs-soroban-sdk#913]
_Note: subjective developer ergonomics. low priority_

Contract implementations currently support both contract functions that include an `Env` as their first argument, and ones that don't. In practice, I can't think of any meaningful contract functions that don't interact with `Env`.  

This gets very verbose and (subjectively) I find it an eyesore. Especially when defining and documenting contract traits, such as the [token interface](https://soroban.stellar.org/docs/reference/interfaces/token-interface), which we're about to do a lot of for ecosystem standard. 

Is this important enough to change? And if so, how? Maybe place the `Env` inside the struct? (cc @leighmcculloch)

**Comment by leighmcculloch** [comment]
I agree, it's a bit of a shame the `Env` is littered through the functions. A side-effect is that clients for contracts and contracts themselves don't adhere to the same interface. If we removed the `Env` from the function parameters they would match and we could do things like have clients implement the trait of a contract. There's no big wins from that, so I wouldn't jump at doing something just for that reason, but it would just look nice/convenience/predictable.

We could move the `Env` into the struct for the contract. It would require either the developer to manually specify it on their struct, or for the developer to manually use a new macro on their struct. Neither would be very ergononomic though, because the `contractimpl` macro would have no way of checking that `Env` existed on the struct. `contractimpl` cares about this because it generates code that instantiates the contract struct, and that generated code would inject the `Env`. If a developer left `Env` off or didn't use the extra macro, the compiler will probably display cryptic messages saying an `Env` is missing somewhere, but the 'where' wouldn't be clear because it is being generated by macro generated code not visible to the developer.

Does anyone have alternative ideas?

**Comment by dmkozh** [comment]
Would a good old global variable work? Semantically, `Env` is just a namespace for a bunch of functions (and it should be stripped in Wasm builds). I know that it resolves to a stateful variable when we're building tests, but probably there is some way to work around that.  So the end result would look like `soroban_sdk::ENV.do_stuff()`.

Another spin on that would be to distribute Env as class with a collection of static methods, that would use the global implementation internally (so that we don't need to expose the global variable to the contracts at all), so calls would look like `soroban_sdk::Env::do_stuff()`. 

Adding `Env` to struct might be simpler implementation-wise, but that doesn't make much sense semantically - there aren't many `Env`s out there, there is just one set of host fns instead.

**Comment by ozgunozerk** [comment]
> Does anyone have alternative ideas?

Here is a very straight-forward solution that should not change the architecture dramatically:
what if we remove (hide) `Env` variable from `#[contractimpl]` functions, and the macro itself will append this variable as the first parameter to the methods that it is encapsulating. 

In other words, after the `#[contractimpl]` is expanded, it will be the same (there will be `Env` variable as the first parameter). But it won't be the developers responsibility to put `Env` into the functions, and the macro will put it. 

One downside would be, using the parameter `e` would be seen *magical* and *confusing* for the ones who aren't familiar with this behavior. But the same problem also applies to moving `Env` to the `struct` -> using the env directly as `e` would be confusing. 

Imo, the benefits (better DevX) will outweigh the initial confusion.

If this idea sounds good, I can even volunteer to implement it.

**Comment by dmkozh** [comment]
> In other words, after the #[contractimpl] is expanded, it will be the same (there will be Env variable as the first parameter). But it won't be the developers responsibility to put Env into the functions, and the macro will put it.

Given that Soroban SDK has been around for a while, I think it would be safer to introduce a new macro if we were to do that (something like `[contractimpl_env]`). I don't have too strong opinion about this feature otherwise; it seems like it does reduce boilerplate, but it's also not very pretty due to introducing a magic variable.

**Comment by leighmcculloch** [comment]
> One downside would be, using the parameter `e` would be seen _magical_ and _confusing_ for the ones who aren't familiar with this behavior.

The surprising side of this for a code reader is the big tradeoff with this approach. I agree that moving the Env into the struct is also surprising, but both appear in different ways.

If the parameter is inserted, it will be surprising to a dev if they try and call the function internally. They'll look at the code and see no Env, but at the call sites the Env will need to be there.

For example:

```rust
#[contract]
pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn balance(a: Address) -> i128 {
        // ...
        e.storage()...
        // ...
    }

    pub fn transfer(from: Address, to: Address, amount: i128) {
        // ...
        let from_bal = Self::balance(e, &from);
        // ...
    }
}
```

Where-as, if the Env is embedded into the Contract, it is reasonably expected that functions can be added to types via macros and be used elsewhere.

For example:

```rust
#[contract]
pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn balance(&self, a: Address) -> i128 {
        // ...
        self.env().storage()...
        // ...
    }

    pub fn transfer(&self, from: Address, to: Address, amount: i128) {
        // ...
        let from_bal = self.balance(&from);
        // ...
    }
}
```
